oe_content.module

```diff
@@ -106,3 +109,128 @@ function oe_content_entity_base_field_info(EntityTypeInterface $entity_type) {
 function oe_content_locale_translation_projects_alter(&$projects) {
   $projects['oe_content']['info']['interface translation server pattern'] = drupal_get_path('module', 'oe_content') . '/translations/%project-%language.po';
 }
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ *
+ * Add the composite field option to entity reference fields
+ */
+function oe_content_form_field_config_edit_form_alter(&$form, FormStateInterface $form_state): void {
+  /** @var \Drupal\field\FieldConfigInterface $field_config */
+  $field_config = $form_state->getFormObject()->getEntity();
+  $applicable_field_types = ['entity_reference', 'entity_reference_revisions'];
+  if (!in_array($field_config->getType(), $applicable_field_types)) {
+    return;
+  }
+  $form['composite'] = [
+    '#type' => 'radios',
+    '#title' => t('Composite field'),
+    '#description' => t('Making this field a composite field will delete any referenced entity if the referencing entity is deleted (only if the referenced entity is not referenced by some other entity).'),
+    '#default_value' => $field_config->getThirdPartySetting('oe_content', 'composite') ? 1 : 0,
+    '#options' => [
+      1 => t('Enabled'),
+      0 => t('Disabled'),
+    ],
+  ];
+
+  $form['#entity_builders'][] = 'oe_content_form_field_config_form_builder';
+
+}
+
+/**
+ * Entity form builder for the field_config form.
+ *
+ * Saves the composite field settings into the field config settings.
+ *
+ * @param string $entity_type
+ *   The name of the entity type.
+ * @param \Drupal\field\FieldConfigInterface $field_config
+ *   The field config.
+ * @param array $form
+ *   The form array.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The form state.
+ */
+function oe_content_form_field_config_form_builder(string $entity_type, FieldConfigInterface $field_config, array &$form, FormStateInterface $form_state): void {
+  if ($form_state->getValue('composite')) {
+    $field_config->setThirdPartySetting('oe_content', 'composite', TRUE);
+  }
+  else {
+    $field_config->setThirdPartySetting('oe_content', 'composite', FALSE);
+  }
+}
+
+/**
+ * Implements hook_entity_delete().
+ *
+ * Delete any composite referenced entities when deleting an entity.
+ */
+function oe_content_entity_delete(EntityInterface $entity): void {
+  if (!$entity instanceof \Drupal\Core\Entity\FieldableEntityInterface) {
+    return;
+  }
+  /** @var \Drupal\Core\Field\FieldDefinitionInterface[] $field_definitions */
+  $field_definitions = \Drupal::service('entity_field.manager')->getFieldDefinitions($entity->getEntityTypeId(), $entity->bundle());
```

Comment:
Let's move all this logic that looks for fields and does the delete and checks if it's not referennced already do a standalone service. Let's have something like: 

`CompositeReferenceFieldManager::onDelete()` getting the field value. So you check in the hook if it's the correct field type, then you defer to the service. And in the service you get the references, check them and delete them.

For the method that checks if we can delete, let's name it and make it generic: `getReferencingEntities()` and it should return info about the referencing entities. We might use this elsewhere as well, or someone would.

Reply:
Done, please check if the service is what you where envisioning.